Split EMF log with larger than 100 buckets.

zzhlogin opened the pull request:
**Description:** <Describe what has changed.>
In Application Signals, we utilize Base2 Exponential Bucket Histogram to aggregate and send latency data, with a default max number of buckets 160. In EMF exporter, these buckets are mapped to "[Target members](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Embedded_Metric_Format_Specification.html#CloudWatch_Embedded_Metric_Format_Specification_structure_target)" in EMF log entries. 
However, CloudWatch EMF logs impose a limit of 100 target members, beyond which EMF processors will mark the record as `invalid`, resulting in missing metrics and customer-facing errors reported via the EMFValidationErrors metric.

In this PR, we split histograms to two sub EMF logs with the following change:
1. Add an extra attribute `metricIndex` to `groupedMetricMetadata` : Current EMF exporter aggregate incoming metrics into groupedMetrics before converting to log events, where the groupKey is generated based on the groupedMetricMetadata including: metric namespace, timestamp, log group name, etc. After splitting, the two new metrics will share exactly the same key. Adding an extra metric metadata for key generation can prevent the second metric from dropping.
2. If the total buckets exceed 100, the exponential histogram metric are split into into multiple data points as needed,
each containing a maximum of 100 buckets, to comply with CloudWatch EMF log constraints.
For each split data point:
- Min and Max values are recalculated based on the bucket boundary within that specific split.
- Sum is only assigned to the first split to ensure the total sum of the datapoints after aggregation is correct.
- Count is accumulated based on the bucket counts within each split.
 
**Testing:** <Describe what testing was performed and which tests were added.>
The change is tested by generating traffic with more than 100 buckets, and the emf log with larger than 100 values are eliminated after the change:
![Screenshot 2024-10-18 at 3 04 45 PM](https://github.com/user-attachments/assets/8c2aac01-6f34-44e9-bcaa-3f7b1f16fc7d)

Also in E2E test, logging the Min/Max/Count to confirm the behaviour is expected.

zzhlogin committed:
Add test log.

zzhlogin committed:
code clean up.

bjrara commented on exporter/awsemfexporter/datapoint_test.go:
```diff
@@ -939,6 +968,89 @@ func TestCalculateDeltaDatapoints_ExponentialHistogramDataPointSlice(t *testing.
+					Sum: 1000, Count: 2612, Min: -9e+17, Max: 5.49755813888e+11},
+				labels: map[string]string{"label1": "value1"},
+			},
+			expectedDatapoint2: dataPoint{
```
You said `the second data point contains the last 100 buckets, while the first data point includes the remaining buckets` but it doesn't seem to be the case?

zzhlogin replied:
Yes, the reason here is because return the second split as the first datapoint here, so the order switched. I have updated the code  to set the first bucket has in max on 100 values, and return the datapoint in the right order to avoid confusion.

bjrara reviewed (commented):
Thanks for the contribution. Some high level comments:

1. Intuitively, I think that the first split would contain as many datapoints as possible. Let's say in the future, if the EMF exporter is gonna support "transactional exporting", i.e. the second batch will only export if the first one succeeds. In this case, we want the majority to be exported in the first to reduce the risk of failures in the second. Switching the number should not impact how many data you would traverse in the splitting logics. As for now, it's not a big concern because each exporting is independent.
2. Shall we also include `histogramDataPointSlice`?
3. The test cases don't seem to cover enough use cases, e.g. with positive empty, with negative empty, with 0 empty, with non-splitting validation when the total size is smaller than 100. Please add more cases to verify the logics. 
4. Regarding the only one test case you added, having a large array base for result validation is good, but how did we confirm that the `expectedDatapoint` is what we really want? It's hard to judge the accuracy.

zzhlogin commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,42 +213,94 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
-	arrayCounts := []float64{}
 	var bucketBegin float64
 	var bucketEnd float64
+	firstDataPointCount := 0
```
Yes, create a `dataPointSplit` construct with `newDataPointSplit` function, which can be initialized for each split.

zzhlogin commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,42 +213,94 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
 }
 
 // CalculateDeltaDatapoints retrieves the ExponentialHistogramDataPoint at the given index.
+// If the total buckets exceed 100, the exponential histogram metric are split into two data points,
```
Added

zzhlogin commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,42 +213,94 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
+	currentLength := 0
+	firstDataPointLength := totalBucketLen - 100
+	if firstDataPointLength < 0 {
+		firstDataPointLength = 0
```
If firstDataPointLength is 0, we don't need to split the bucket, but we still need to iterating the buckets as we need to calculate the midValue for each bucket. I have updated the code to set the first bucket has in max on 100 values, so we will never reach to second bucket if we have less than 100 buckets.

zzhlogin commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -80,6 +80,7 @@ type groupedMetricMetadata struct {
 	logGroup                   string
 	logStream                  string
 	metricDataType             pmetric.MetricType
+	metricIndex                int
```
Updated to use `batchIndex`

zzhlogin commented:
> Thanks for the contribution. Some high level comments:
> 
> 1. Intuitively, I think that the first split would contain as many datapoints as possible. Let's say in the future, if the EMF exporter is gonna support "transactional exporting", i.e. the second batch will only export if the first one succeeds. In this case, we want the majority to be exported in the first to reduce the risk of failures in the second. Switching the number should not impact how many data you would traverse in the splitting logics. As for now, it's not a big concern because each exporting is independent.

Updated the code to set the first bucket has in max on 100 values, and return the datapoint in the right order to avoid confusion.

> 2. Shall we also include `histogramDataPointSlice`?

The `histogramDataPointSlice.CalculateDeltaDatapoints` only return [metrics Stats](https://github.com/amazon-contributing/opentelemetry-collector-contrib/blob/9ae2897e22a496f5527d55e13af60bbceae70a4e/exporter/awsemfexporter/datapoint.go#L191), including `count`, `sum`, `max`, `min`, there is no `values` and `counts` returned, the members won't exceed 100.

> 3. The test cases don't seem to cover enough use cases, e.g. with positive empty, with negative empty, with 0 empty, with non-splitting validation when the total size is smaller than 100. Please add more cases to verify the logics.

Added all these test cases. (The verification principle follows the explanation in 4. below)

> 4. Regarding the only one test case you added, having a large array base for result validation is good, but how did we confirm that the `expectedDatapoint` is what we really want? It's hard to judge the accuracy.

In the test case, we generate in total of 121 different buckets, with different number of values counts. The total count is 3662 (the result is calculated locally by iterating over all buckets), and the sum is hard coded to 1000 (to verify we directly use metric sum in the first split). 
- For the first split: `Max` is retrieved from `Metric.Max` value, and the `Min` is retrieved from the 100th bucket lower boundary (I confirmed by locally print out the 100th bucket and retrieve the value).  It has exactly 100 values it is checked in expectedDatapoint1: value and Counts.
-  Similar For second split, `Min` is retrieved from `Metric.Min` value, `Max` is the same as first split's `Min`. It has exactly 21 values it is checked in expectedDatapoint2: value and Counts.

zzhlogin committed:
Address comments.

jefchien commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -105,6 +105,28 @@ type summaryMetricEntry struct {
+	maxBuckets int
+}
+
+//func newDataPointSplit(count int, length int, bucketMin float64, bucketMax float64, maxBuckets int) *dataPointSplit {
```
nit: Why is this commented out?

zzhlogin replied:
Sorry, missed when doing code clean up. Deleted

jefchien commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -273,25 +375,72 @@ func (dps exponentialHistogramDataPointSlice) CalculateDeltaDatapoints(idx int,
+	})
+
+	// Add first data point with the remaining elements
+	if totalBucketLen > 100 {
```
Seems like this should be 
```suggestion
	if totalBucketLen > splitThreshold {
```

zzhlogin replied:
Updated

jefchien commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -105,6 +105,28 @@ type summaryMetricEntry struct {
 	count uint64
 }
 
+type dataPointSplit struct {
+	values     []float64
+	counts     []float64
+	count      int
+	length     int
+	bucketMin  float64
+	bucketMax  float64
+	maxBuckets int
+}
```
nit: Why do we need a new struct? Can't we reuse the `cWMetricHistogram`? It seems like most of these (aside from `length` and `maxBuckets`) map 1:1 with the fields in the `cWMetricHistogram` which is the end result. Could have the struct be something like

```suggestion
type dataPointSplit struct {
	cWMetricHistogram
	length     int
	capacity int // seems like maxBuckets is being used in this manner similar to `len` and `cap`
}
```

zzhlogin replied:
Updated with the struct suggested.

jefchien commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,42 +235,122 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
+//   - Min: From original metric.
+//   - Sum: 0.
+//   - Count: Overall count - first split count.
 func (dps exponentialHistogramDataPointSlice) CalculateDeltaDatapoints(idx int, _ string, _ bool, _ *emfCalculators) ([]dataPoint, bool) {
```
I know for our usecase, we only expect a max of 160 buckets, but if we get more than 200, this still won't work. Is there a reason why we're restricting ourselves to 2 datapoint splits? Can't we support N datapoint splits?

zzhlogin replied:
You are correct, I was checking the otel upstream Exponential Bucket Histogram Aggregation has [Maximum number of buckets 160](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#base2-exponential-bucket-histogram-aggregation). But it turns out the 160 is the maximum number in each of the positive and negative ranges, so the total number can be more than 300. I have changed to support N datapoint splits.

jefchien commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,42 +235,122 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
 	bucketBegin = 0
 	bucketEnd = 0
-	for i := 0; i < positiveBucketCounts.Len(); i++ {
+	for i := positiveBucketCounts.Len() - 1; i >= 0; i-- {
```
Did we just reverse this so the max value is in the first split?

zzhlogin replied:
Yes,  the positiveBucket range is from low to high, I'm reverting it so that we can always assign higher values to front datapoints

jefchien commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -273,25 +375,72 @@ func (dps exponentialHistogramDataPointSlice) CalculateDeltaDatapoints(idx int,
+	//fmt.Println("secondSplit", secondSplit)
+
+	var datapoints []dataPoint
+	// Add second data point (last 100 elements or fewer)
```
Isn't this adding the first datapoint?

zzhlogin replied:
The code here was assigning the datapoint and reversely. I have updated to code to add N datapoint sequentially.

jefchien commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,42 +235,122 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
+				}
+			}
+		} else if count > 0 {
+			if currentLength == firstSplit.maxBuckets && currentLength != 0 {
```
Why do we need the `currentLength`? Shouldn't checking `firstSplit.length` against `firstSplit.maxBuckets` be enough?

zzhlogin replied:
Make sense, I have updated the PR.

jefchien commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,42 +235,122 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
+			firstSplit.counts = append(firstSplit.counts, float64(count))
+			firstSplit.length++
+			currentLength++
+			if firstSplit.maxBuckets < totalBucketLen {
```
So if the total buckets is > 100, then the `firstSplit` will have a `count` of 0?

zzhlogin replied:
I have updated to code to add N datapoint sequentially. each datapoint will have actual count.

jefchien commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,42 +235,122 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
+// 2. Second split datapoint:
+//   - Max: First bucket’s bucketEnd in the second split.
+//   - Min: From original metric.
+//   - Sum: 0.
```
Why is the `Sum` set to 0 in this case? Shouldn't the two datapoints be essentially separate?

zzhlogin replied:
The reason why we are not calculating Sum base on bucket datapoint: 
the raw Sum is using the raw metric data. Since the otel collector doesn’t have visibility of metric raw data,  if we recalculate the sum values for the split datapoint, we can only use the approximate value (mid value for each bucket) which might have an offset of the real value


In this case:
I'm only assigning actual `Sum` to the first split, all the following split will have sum equals to 0. Then, after emf log aggregation, the total sum of the datapoints is exactly same before split.

zzhlogin committed:
Split to multiple datapoints according to input size.

zzhlogin committed:
Update comments.

zzhlogin committed:
Fix counter part.

bjrara commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,85 +219,184 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
-		arrayValues = append(arrayValues, 0)
-		arrayCounts = append(arrayCounts, float64(metric.ZeroCount()))
-	}
+		if totalBucketLen-currentBucketIndex < splitThreshold {
```
`<=` or `<`? Should 100 buckets qualify here?

zzhlogin replied:
if `totalBucketLen-currentBucketIndex == splitThreshold` which is 100, we don't need to change the capacity as it is initialized with 100. We only need to override it when the remaining buckets is less than 100.

bjrara replied:
Can we put `capacity` assignment BEFORE initialization, so that `capacity` becomes immutable, and we never bother changing its value after initialization.

bjrara commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,85 +219,184 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
+					if bucketBegin < bucketEnd {
+						split.cWMetricHistogram.Max = bucketEnd
+					} else {
+						split.cWMetricHistogram.Max = bucketBegin
```
Sorry that I'm struggling to understand the logics here. Under what condition will this be hit?

zzhlogin replied:
Please check the explanation below.

bjrara commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,85 +219,184 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
+				split.cWMetricHistogram.Count += count
+				if split.length == 1 && currentBucketIndex != 0 {
+					if bucketBegin < bucketEnd {
+						split.cWMetricHistogram.Max = bucketEnd
```
I'm wondering if in any case, the `bucketEnd` will be larger than metric max value? Same to min.

zzhlogin replied:
For positive buckets, bucketEnd will always be larger than bucketBegin, so we take bucketEnd as the max value. 

For negative buckets, bucketEnd (`-math.Pow(base, float64(index))`) will always be smaller than bucketBegin(`-math.Pow(base, float64(index+1))`), in this case, we need to take bucketBegin as the max value.

Instead of separating logic for both positive and negative buckets, I'm directly compare the value here to make sure we always take the larger value as the Max. (Same for both positive and negative buckets calculation)

zzhlogin replied:
After discussion, I remove the comparison and directly use the exact value to assign.

jefchien commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,85 +219,184 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
-			bucketBegin = math.Pow(base, float64(index))
-		} else {
-			bucketBegin = bucketEnd
+	if totalBucketLen == 0 {
```
Is an exponential histogram without buckets valid?

zzhlogin replied:
Yes, if there is no bucket valid in the metric, we return a datapoint of empty array instead of return an empty `[]dataPoint`.

zzhlogin committed:
Reformate code.

bjrara commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,85 +219,189 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
+		}
+
+		// Set mid-point of positive buckets in values/counts array.
+		currentBucketIndex, currentPositiveIndex = iteratePositiveBuckets(&split, metric, currentBucketIndex, currentPositiveIndex, totalBucketLen)
```
`iterate` doesn't seem to be a good name. Sometimes, the iteration won't even happen if `split.length >= split.capacity`. Will sth. like `collectDatapointsWithPositiveBuckets` be more accurate?

zzhlogin replied:
Updated

bjrara commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,85 +219,189 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
 		if count > 0 {
-			arrayValues = append(arrayValues, metricVal)
-			arrayCounts = append(arrayCounts, float64(count))
+			split.cWMetricHistogram.Values = append(split.cWMetricHistogram.Values, metricVal)
```
Can we wrap all the assignments in to a function in `split`?

zzhlogin replied:
Updated

bjrara commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,85 +219,189 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
+	bucketBegin := 0.0
+	bucketEnd := 0.0
+
+	for split.length < split.capacity && currentPositiveIndex >= 0 {
```
Can we put do an additional check on `split.length < split.capacity && currentPositiveIndex >= 0` at the beginning? Seems if the condition check fails on start, all the remaining logics will become redundant, and we can exit early in that case.

bjrara replied:
Same to the other `iterate`.

bjrara replied:
Can we wrap `split.length < split.capacity` as a method? It will be more readable as sth. like `!split.isFull()`.

zzhlogin replied:
Updated

zzhlogin committed:
reformate code.

jefchien commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,85 +238,195 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
+			capacity: capacity,
+		}
+
+		// Set mid-point of positive buckets in values/counts array.
+		currentBucketIndex, currentPositiveIndex = collectDatapointsWithPositiveBuckets(&split, metric, currentBucketIndex, currentPositiveIndex)
+		// Set count of zero bucket in values/counts array.
+		currentBucketIndex, currentZeroIndex = collectDatapointsWithZeroBuckets(&split, metric, currentBucketIndex, currentZeroIndex)
+		// Set mid-point of negative buckets in values/counts array.
+		currentBucketIndex, currentNegativeIndex = collectDatapointsWithNegativeBuckets(&split, metric, currentBucketIndex, currentNegativeIndex)
```
So is the idea that if the split becomes full at some point, all of these will be no-ops?

zzhlogin replied:
Yes, if the split becomes full, these methods will be no-ops, and it will go directly into next loop to create a new split

jefchien commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -80,6 +80,7 @@ type groupedMetricMetadata struct {
 	logGroup                   string
 	logStream                  string
 	metricDataType             pmetric.MetricType
+	batchIndex                 int
```
What is this used for?

zzhlogin replied:
EMF exporter aggregate incoming metrics into groupedMetrics, where the groupKey is generated based on the groupedMetricMetadata. After splitting, the two new split will share exactly the same key. Here I add an extra metric metadata for key generation, this new entry won't be added into emf log, and it can prevent the second metric from dropping.

jefchien commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,85 +238,195 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
+		if split.length == 1 {
+			split.setMax(0)
+		}
+		if split.length == split.capacity {
```
nit: 
```suggestion
		if !split.isNotFull() {
```

jefchien commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,85 +238,195 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
+			capacity: capacity,
+		}
+
+		// Set mid-point of positive buckets in values/counts array.
```
nit: Comment seems outdated.

jefchien commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,85 +238,195 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
 	metric := dps.ExponentialHistogramDataPointSlice.At(idx)
 
+	const splitThreshold = 100
+	var currentBucketIndex = 0
+	var currentPositiveIndex = metric.Positive().BucketCounts().Len() - 1
+	var currentZeroIndex = 0
+	var currentNegativeIndex = 0
```
nit: Don't need to declare var in front of each of these.
```suggestion
	currentBucketIndex := 0
	currentPositiveIndex := metric.Positive().BucketCounts().Len() - 1
	currentZeroIndex := 0
	currentNegativeIndex := 0
```

zzhlogin committed:
Address lint.

vastin commented on exporter/awsemfexporter/datapoint.go:
```diff
@@ -213,85 +238,195 @@ func (dps histogramDataPointSlice) IsStaleNaNInf(i int) (bool, pcommon.Map) {
+	return currentBucketIndex, currentPositiveIndex
+}
+
+func collectDatapointsWithZeroBuckets(split *dataPointSplit, metric pmetric.ExponentialHistogramDataPoint, currentBucketIndex int, currentZeroIndex int) (int, int) {
```
nit: There is only one zero bucket.  Probably rename function to `collectDatapointsWithZeroBucket`.

vastin reviewed (approved):
LGTM

zzhlogin committed:
Address comments.

zzhlogin committed:
eliminate zero splits.
